src/narrowphase/gjk.cpp

```diff
@@ -689,8 +690,29 @@ GJK::Status GJK::evaluate(const MinkowskiDiff& shape_, const Vec3f& guess,
     }
 
     // check C: when the new support point is close to the sub-simplex where the ray point lies, stop (as the new simplex again is degenerated)
-    FCL_REAL duality_gap = std::sqrt(2 * ray.dot(ray - w));
-    if(iterations > 0 && duality_gap - tolerance * rl <= 0)
+    // -- CURRENT HPPFCL CRITERION: ||ray|| - ||xstar|| <= rl - alpha (where rl = ||ray||, xstar is unknown optimal solution)
+    /* alpha = std::max(alpha, omega); */
+    /* FCL_REAL diff = rl - alpha; */ 
+    /* bool cv_check_passed = diff - tolerance * rl <= 0; */
+
+    // -- DUALITY GAP: ||ray - xstar|| <= sqrt(2 * ray.dot(ray - w))
+    /* FCL_REAL diff = std::sqrt(2 * ray.dot(ray - w)); */
+    /* bool cv_check_passed = diff - tolerance * rl <= 0; */
+    // ---- Squared version:
+    FCL_REAL diff = 2 * ray.dot(ray - w);
+    bool cv_check_passed = diff - tolerance * tolerance * rl * rl <= 0;
+
+    // -- IMPROVED DUALITY GAP: ||ray - xstar|| <= sqrt(rl*rl - alpha*alpha)
+    /* alpha = std::max(alpha, omega); */
+    /* FCL_REAL diff = std::sqrt(rl * rl - alpha * alpha); */ 
+    /* bool cv_check_passed = diff - tolerance * rl <= 0; */
+    // ---- Squared version:
+    /* alpha = std::max(alpha, omega); */
+    /* FCL_REAL diff = rl * rl - alpha * alpha; */ 
```

Comment:
The **improved** duality-gap also passed all tests, **except** for obb and octree and gjk. But it passes the collision test. Weird.
When using square root version (commented just above), it fails everything. Again, it should be equivalent so the tests should return the same passes/fails. I don't get it.